### ISSUE TITLE
Enhancement: `RuleTester#run` - Infer `MessageIds` and `Options` type parameter from `rule` parameter

### ISSUE THREAD

**Issue by marcalexiei** [typescript-eslint/typescript-eslint#10321]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

rule-tester

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Right now `RuleTester#run` method accepts `MessageIds` and `Options` as explicit type parameters:
https://github.com/typescript-eslint/typescript-eslint/blob/90617a651c3a8b244cdcc2a618aa999b7643ecce/packages/rule-tester/src/RuleTester.ts#L479-L486

My proposed enhancement is to make the two types inferred from the rule parameter using a "type extractor": 

```ts
import type { RuleModule } from '@typescript-eslint/utils/ts-eslint';

type GetRuleInternalTypes<TRule> =
  TRule extends RuleModule<infer MessageIds, infer Options>
    ? {
        messageIds: MessageIds;
        options: Options;
      }
    : {
        messageIds: string;
        options: Array<unknown>;
      };

```

to infer the rule types from the `rule` parameter provided to `run` method.
This way the `run` method signature will become this:

```ts
run<
    MessageIds extends string = never,
    Options extends readonly unknown[] = never,
    TRule extends RuleModule<string, unknown[]> = RuleModule<string, unknown[]>,
    TRuleInternalTypes extends
      GetRuleInternalTypes<TRule> = GetRuleInternalTypes<TRule>,
    ActualMessageIDs extends string = [MessageIds] extends [never]
      ? TRuleInternalTypes['messageIds']
      : MessageIds,
    ActualOptions extends readonly unknown[] = [Options] extends [never]
      ? TRuleInternalTypes['options']
      : Options,
  >(
    ruleName: string,
    rule: TRule,
    test: RunTests<NoInfer<ActualMessageIDs>, NoInfer<ActualOptions>>,
  ): void;
```

Using this approach the signature of the function is backward compatible:
- If someone was explicating type parameters, their values will take precedence over the one inferred from the `rule` parameter.
- if the rule is not a `RuleModule` it will use the default parameters
- if the rule extends `RuleModule` the messageIds and options will be inferred (✨ enhancement!)

<details>
<summary>Original proposal (surely a breaking change)</summary>

```ts
   /**
     * Adds a new rule test to execute.
     */
    run<
      TRule extends RuleModule<string, unknown[]>,
      TData extends GetRuleInternalTypes<TRule> = GetRuleInternalTypes<TRule>
    >(
      ruleName: string,
      rule: TRule,
      test: RunTests<TData['messageIds'], TData['options']>,
    ): void;
```

---

IMHO I found this change very useful when writing rules test since all `messageId` and `options` are suggested via IDE autocomplete.

Since this a breaking change, a new `run` function (I have no name in mind at the moment 😅) for `RuleTester` could be created to avoid a breaking change release or waiting for the next major release.

</details>

---

If you are interested with the change I'm more than willing to open a PR with the solution that is more suitable to you!

### Additional Info

Inspiration for this proposal comes from https://github.com/un-ts/eslint-plugin-import-x/pull/176#discussion_r1835998981

**Comment by SukkaW** [comment]
@marcalexiei Is it possible to append (instead of prepend) the `TRule` generic? This way we can prevent a huge breaking change with massive migration.

**Comment by marcalexiei** [comment]
After a little tackling I found a backward compatibility to avoid breaking changes. I'll update the description shortly.

**Comment by kirkwaiblinger** [comment]
This sounds like a great suggestion! Opening to PRs.

**Comment by bradzacher** [comment]
I'm not quite sure I understand - how is your suggestion different to the current signature?
Couldn't we just mark the `tests` parameter as `test: RunTests<NoInfer<MessageIDs>, NoInfer<Options>>` and have the same behaviour?

**Comment by marcalexiei** [comment]
> TS will infer the MessageIds and Options type parameters from the function arguments already - as evidenced from our codebase where we rely on this behaviour.
> how is your suggestion different to the current signature?

You are right, I just realised it when starting working on the PR.
Changing the signature with my proposal only complicates it.

TL;DR;
All tests were wrapped in an helper function that was causing wrong type inference and I haven't checked this behaviour on a minimal repository.
Clearly I made an oversight 😅, thanks for pointing it out.

<details>
<summary>Details about my wrong assumption of wrong inference</summary>

On [eslint-plugin-import-x](https://github.com/un-ts/eslint-plugin-import-x) all test cases are wrapped in a [`test` function which is returning `InvalidTestCase` or a `ValidTestCase`](https://github.com/un-ts/eslint-plugin-import-x/blob/26cdedf91f846d1bb5bc565b0a746d849ad10c78/test/utils.ts#L80-L84) without any generic type parameter.

That was causing the wrong parameter on `RuleTester#run` producing something like:

```ts
(method) RuleTester.run<
  MessageId, 
  any
>(
  ruleName: string, 
  rule: RuleModule<MessageId, any, unknown, RuleListener>, 
  test: RunTests<MessageId, any>
): void
```

instead of 
```ts
(method) RuleTester.run<
  MessageId, 
  [(Options | undefined)?]
>(
  ruleName: string, 
  rule: RuleModule<MessageId, [(Options | undefined)?], unknown, RuleListener>,
  test: RunTests<...>,
): void
```

While on code without the usage of the helper function everything works correctly:

```ts
ruleTester.run("prefer-default-export", rule, {
  valid: [
    { code: "", options: [{ target: "any" }] },
    { code: "", options: [{ target: "a2y" }] }, // error
  ],
  invalid: [
    {
      code: "asd",
      options: [{ target: "single" }],
      errors: [{ messageId: "invalidID" }], // error
    },
  ],
});
```
</details>

---

> If you're meaning that you don't want the test parameter to influence the generic -- then couldn't we just mark the tests parameter as test: RunTests<NoInfer<MessageIds>, NoInfer<Options>> and have the same behaviour?

Yeah, that surely solves the wrong infer I was experiencing. Is it a suitable change on your side?

**Comment by bradzacher** [comment]
More than happy if you want to add `NoInfer` to the tests param - that would be a good win!
